ENH: add cumulative_simpson integration to scipy.integrate

nprav opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
N/A

#### What does this implement/fix?

The cumulative integration and differentiation of numerical samples is a common operation in the engineering domains. 
Scipy currently only provides the integrate.cumulative_trapezoid function, but this is insufficiently accurate for engineering purposes. The integrate.simpson method is available but only as a complete integral, not a cumulative integral.  This PR adds an implementation for the cumulative integration of an array of samples using the Simpson's 1/3 rule ([wiki link](https://en.wikipedia.org/wiki/Simpson%27s_rule)).

CAE softwares such as ANSYS/LS-DYNA provide simpson's integration as a default processing option. Adding this to Scipy.integrate will help expand usage in the engineering domains.


#### Additional information
<!--Any additional information you think is important.-->
Refer to this jupyter notebook: [Cumulative Simpson derivation](https://github.com/nprav/misc/blob/main/cumulative_simpson/Cumulative%20Simpson%20algorithm%20derivation.ipynb) for further information.

tupui reviewed (commented):
Thank you for the PR @nprav!

Before we start any review, we need to get the approval for inclusion in SciPy. The process involves sending an email to the dev mailing list. Could you do that?

https://mail.python.org/mailman3/lists/scipy-dev.python.org/

MatteoRaso commented on scipy/integrate/_quadrature.py:
```diff
@@ -1304,15 +1514,15 @@ def qmc_quad(func, a, b, *, n_estimates=8, n_points=1024, qrng=None,
     >>> res = qmc_quad(func, a, b, n_estimates=n_estimates, qrng=qrng)
     >>> res.integral, res.standard_error
-    (0.00018443143881633162, 4.709434153066518e-08)
+    (0.00018441088533413305, 1.1255608140911588e-07)
 
     A two-sided, 99% confidence interval for the integral may be estimated
     as:
 
     >>> t = stats.t(df=n_estimates-1, loc=res.integral,
     ...             scale=res.standard_error)
     >>> t.interval(0.99)
-    (0.00018426663295474533, 0.0001845962446779179)
+    (0.00018401699720722663, 0.00018480477346103947)
```
This change doesn't seem to have anything to do with what the PR's about. If you changed this because the output given in the example is wrong, it would be best to deal with that in a separate PR.

tupui commented:
@MatteoRaso Thank you for starting a review but please wait until there is a decision from a maintainer. We don't want people to spend time on something that could in the end be rejected.

lorentzenchr commented on scipy/integrate/_quadrature.py:
```diff
@@ -662,6 +664,216 @@ def simpson(y, x=None, dx=1.0, axis=-1, even='avg'):
+    y = y.swapaxes(0, axis)
+    dx = dx.swapaxes(0, axis)
+
+    sub_integrals_fwd = dx[:-1] / 3 * \
+        (5 * y[:-2] / 4 + 2 * y[1:-1] - y[2:] / 4)
+    sub_integrals_bwd = dx[:-1] / 3 * \
+        (-y[:-2] / 4 + 2 * y[1:-1] + 5 * y[2:] / 4)
```
It would be nice to know where these formulae come from.
For instance, an explanation like: The first is the Simpson rule for the integration part from a to a+h, the second from a+h to a+2h=b (but note that wikipedia uses a different h, I think). Adding both parts gives the famous Simpson rule, see wikipedia....

nprav replied:
Thank you for the review @lorentzenchr !

Just to clarify, are you requesting I add this information as a comment above line 674? Or, can I add it elsewhere such as the docstring of the cumulative_simpson function?

Wikipedia uses the same 'h'. It is defined as the 'step size': (b-a)/2
![image](https://user-images.githubusercontent.com/7060774/227790626-26a1cc4e-cd20-40b7-a619-d92791d3e6af.png)

But, Wikipedia does not list the solution for unequal intervals. This is derived in this notebook: [Cumulative Simpson derivation](https://github.com/nprav/misc/blob/main/cumulative_simpson/Cumulative%20Simpson%20algorithm%20derivation.ipynb). Any suggestion for how I can reference this (if needed)?

lorentzenchr replied:
A scipy maintainer might want to step in.
My personal order of preference is:
1. A Notes section in the docstring where one can use LaTeX.
2. A comment in the code

For the formula with unequal steps, I don‘t know what‘s best. Is there a paper or book that can be referenced?

nprav replied:
Thanks @lorentzenchr 

My mistake, the solution actually IS available on [wikipedia](https://en.wikipedia.org/wiki/Simpson%27s_rule), but with slightly different notation. I can reference this.
![image](https://user-images.githubusercontent.com/7060774/227803039-3cafcc18-812b-4256-b379-30e0bf68b1d2.png)

The Notes section in the docstring with added references sounds good.
I can then add a single line comment to improve clarity in the code itself.

Does this work?

nprav replied:
Notes and additional line comments have been added to explain formulae and link to Wikipedia details.

lorentzenchr commented on scipy/integrate/_quadrature.py:
```diff
@@ -662,6 +664,216 @@ def simpson(y, x=None, dx=1.0, axis=-1, even='avg'):
+
+def cumulative_simpson(y, x=None, dx=1.0, axis=-1, initial=None):
+    """
+    Cumulatively integrate y(x) using Simpson's 1/3 rule. The integral of the samples
+    at every point is calculated by assuming a quadratic relationship between each point and two
+    other adjacent points.
```
```suggestion
    Cumulatively integrate y(x) using Simpson's 1/3 rule.
    
    The integral of the samples at every point is calculated by assuming a quadratic
    relationship between each point and the two adjacent points.
```

lorentzenchr commented on scipy/integrate/_quadrature.py:
```diff
@@ -662,6 +664,216 @@ def simpson(y, x=None, dx=1.0, axis=-1, even='avg'):
+
+    """
+
+    y = np.asarray(y).astype(np.float64)
```
Why does it need to be cast to float64?

nprav replied:
The dtypes were cast to ensure all inputs (y, x, dx, initial) are of the same type. This way there is no chance of an inaccurate result where a float array is output as an int32. 

For eg in [line 872](https://github.com/scipy/scipy/pull/18151/files#diff-2ff88b8913ad74b9ea3fe1fe02d6d49164855189168b767076af2e94206c3a2cR872), if initial is of dtype int32, and res is dtype float64, then the output of the function will still be int32.

Casting the dtypes from the outset would ensure the function gives consistent results.

What do you think?

nprav replied:
Thanks for the recommendations @MatteoRaso . I have now removed the casting.

I also updated the formulae so that the term coefficients are [floats](https://github.com/scipy/scipy/pull/18151/commits/cbc46b5f7944db8b87c18f919a1d153edcc0f4f0#). This ensures that even if the inputs are all of type `int`, the output will be in `float` (consistent with cumulative_trapezoid).

Earlier i was having an issue with `initial[slice1] += res`. This was giving an error if initial was `np.int32` and res was `np.float64`. This had led me to explicitly casting, but `initial[slice1] = initial[slice1] + res` does the type casting correctly.

lorentzenchr commented on scipy/integrate/_quadrature.py:
```diff
@@ -662,6 +664,216 @@ def simpson(y, x=None, dx=1.0, axis=-1, even='avg'):
+
+    if y.shape[axis] < 3:
+        raise ValueError(
+            "Minimum of 3 points are required along the axis of integration "
```
```suggestion
            "At least 3 points are required along the axis of integration "
```

lorentzenchr commented on scipy/integrate/_quadrature.py:
```diff
@@ -662,6 +664,216 @@ def simpson(y, x=None, dx=1.0, axis=-1, even='avg'):
+        )
+
+    if x is not None:
+        x = np.asarray(x).astype(np.float64)
```
Same here, why the cast?

nprav replied:
See previous comment.

lorentzenchr commented on scipy/integrate/tests/test_quadrature.py:
```diff
@@ -397,3 +397,222 @@ def func(x):
+            ),
+        ],
+    )
+    def test_simpson_exceptions(self, y_func, x, dx, initial, axis, int_y_func):
```
How about adding a test for `y.shape[0] < 3`?

nprav replied:
Agreed, added this test case

lorentzenchr reviewed (commented):
Some review comments. IMHO, it is already looking quite good and a very useful addition to scipy.

andyfaff commented:
A comment from the sidelines. I would expect such a function to have the same functionality as  the existing [`integrate.simpson`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.simpson.html) function, and for its behaviour to eventually reduce to the existing function.

i.e. it should handle end points similarly, and the behaviour of overlapping keywords should be the same, etc.

nprav commented:
Thank you for the comment @andyfaff ! This is an important discussion point because there are currently some differences in the function signatures. Please find the differences and justifications below and let me know what you think.

### Current function signatures:
- simpson(y, x=None, dx=1.0, axis=-1, even='avg')
- cumulative_trapezoid(y, x=None, dx=1.0, axis=-1, initial=None)
- cumulative_simpson(y, x=None, dx=1.0, axis=-1, initial=None)

### Comparisons
**Integrate.simpson(y, x=None, dx=1.0, axis=-1, even="avg")**
- y, x, dx, axis work the same way
- even is not used in cumulative_simpson 

The 'even' keyword relates to what happens if there are an odd number of intervals. Integrate.simpson will necessarily apply a trapezoidal rule for the odd interval. However, this approach violates the assumption of a quadratic relationship. In reality, the final interval can be approximated with a modified formula while maintaining the quadratic assumption. This is described in the [wikipedia article](https://en.wikipedia.org/wiki/Simpson%27s_rule). 
![image](https://user-images.githubusercontent.com/7060774/227846628-c4ea4262-8cb4-496f-81f5-66782a6f187e.png)

It is also depicted with figures here: [Cumulative Simpson Derivation](https://github.com/nprav/misc/blob/main/cumulative_simpson/Cumulative%20Simpson%20algorithm%20derivation.ipynb).

Because of the above, I had implemented cumulative_simpson without the trapezoidal rule logic. This also ensures that when applied to samples with a quadratic relationship, cumulative_simpson will always give the exact solution. On the other hand, integrate.simpson will be inaccurate if there are an odd number of intervals.

I also compared accuracy when integrating oscillatory functions and found cumulative_simpson to be exhibit equal or better accuracy than integrate.simpson (above notebook link).


**cumulative_trapezoid(y, x=None, dx=1.0, axis=-1, initial=None)**
- y, x, dx, axis work the same way. 
- Initial is slightly different
    - In cumulative_trapezoid, initial is simply prepended to the result array.
    - In cumulative_simpson, initial is prepended to the result array AND summed to every value in the result

The second approach is more 'correct' because the 'initial' value is analogous to a constant of integration that must be added to every point in the cumulative integral. Consider the following practical example:
- Let y be sampled velocity data of an oscillating object. We wish to cumulatively integrate to attain the displacement of the object. We know the starting position (initial value) is 10m. Without the initial value input, the displacement data will show oscillations around 0m. With the initial value, the displacement should show oscillations around 10m. 
- But, with the cumulative_trapezoid implementation, the first value would be 10m, and the object would then be oscillating around 0m, which is incorrect. The summing approach would fix this.

andyfaff commented:
I don't have time to go through this PR myself (plus I'm not an expert in this area of the codebase). However, I should reiterate that behaviour for the proposed PR should tally *exactly* with existing behaviour for `simpson` and `cumulative_trapezoid`. 

For example:

> -Initial is slightly different
  -In cumulative_trapezoid, initial is simply prepended to the result array.
  -In cumulative_simpson, initial is prepended to the result array AND summed to every value in the result

The `cumulative_simpson` would not keep with existing behaviour of `cumulative_trapezoid` because it adds `initial` to every value in the result. This would create confusion with users because of the differing behaviour. See https://github.com/scipy/scipy/pull/125 for why `cumulative_trapezoid` has that behaviour.

Similarly I would expect the exact same behaviour for `simpson` + `cumulative_simpson`, down to the behaviour of the `even` keyword being included, which is not what you outlined here. Different behaviour leads to confusion.

However, I do see the benefit of using the Cartwright2017 approach, as you linked to on Wikipedia. It may be that there's benefit to changing the existing code in `simpson` to take that into account, but that would have to be discussed more thoroughly as it has to take into account back-compatibility.

andyfaff commented:
@mdhaber I'ce prepped a fix for #5618 in https://github.com/andyfaff/scipy/tree/simpson. It needs a bit of discussion. E.g.. how do we deal with back compat, do we ignore `even` kwd, add an extra option to `even`, what's that option called, is that option default, etc.

mdhaber commented:
Yup, your recent email reminded me of those issues.

lorentzenchr commented on scipy/integrate/_quadrature.py:
```diff
@@ -662,6 +664,269 @@ def simpson(y, x=None, dx=1.0, axis=-1, even='avg'):
+    # Calculate for all but first subinterval of y.
+    sub_integrals_h2 = dx[:-1] / 3.0 * \
+        (-y[:-2] / 4.0 + 2.0 * y[1:-1] + 5.0 * y[2:] / 4.0)
+    if y.shape[0] % 2 == 0:
```
Are the code differenced from here on between the equal and unequal interval versions? If it’s the same, it may be factored out.

nprav replied:
The equal and unequal versions are processed the same way after the equations are computed.

I have refactored to bring out the common code into a separate function.

lorentzenchr commented on scipy/integrate/_quadrature.py:
```diff
@@ -991,38 +1240,38 @@ def romberg(function, a, b, args=(), tol=1.48e-8, rtol=1.48e-8, show=False,
 #    where k = N // 2
 #
 _builtincoeffs = {
-    1: (1,2,[1,1],-1,12),
-    2: (1,3,[1,4,1],-1,90),
-    3: (3,8,[1,3,3,1],-3,80),
```
I would revert these cosmetic code changes because they are unrelated. I guess this was you IDE.

nprav replied:
Yup, this was unintentional. Reverted the change.

lorentzenchr commented on scipy/integrate/tests/test_quadrature.py:
```diff
@@ -397,3 +397,240 @@ def func(x):
+
+class TestCumulativeSimpson:
+    @pytest.mark.parametrize(
+        "y_func,x,dx,initial,int_y_func",
```
```suggestion
        "y_func, x, dx, initial, int_y_func",
```
This is easier to read. Or even
```suggestion
        ("y_func", "x", "dx", "initial", "int_y_func"),
```
Same on different other occasions.

nprav replied:
Agreed @lorentzenchr , I have updated the parameters per your second suggestion.

lorentzenchr reviewed (approved):
LGTM. I hope this helps the scipy maintainers.

nprav commented:
Hi @tupui ! What else is required to complete/merge this PR?

tupui commented on scipy/integrate/_quadrature.py:
```diff
@@ -662,6 +664,254 @@ def simpson(y, x=None, dx=1.0, axis=-1, even='avg'):
+    return res.swapaxes(0, axis)
+
+
+def cumulative_simpson(y, x=None, dx=1.0, axis=-1, initial=None):
```
Use some keyword only to simplify future maintenance.
```suggestion
def cumulative_simpson(y, *, x=None, dx=1.0, axis=-1, initial=None):
```

nprav replied:
Incorporated in [be6e1d7](https://github.com/scipy/scipy/pull/18151/commits/be6e1d75e7863d46c297ad24a9929392873dcf3a)

tupui commented on scipy/integrate/_quadrature.py:
```diff
@@ -662,6 +664,254 @@ def simpson(y, x=None, dx=1.0, axis=-1, even='avg'):
+    Notes
+    -----
+
+    .. versionadded:: 1.11.0
```
I am afraid it won't make it to 1.11.0 as we are branching in less than 2 weeks.

nprav replied:
Updated to 1.12.0

tupui commented on scipy/integrate/_quadrature.py:
```diff
@@ -662,6 +664,254 @@ def simpson(y, x=None, dx=1.0, axis=-1, even='avg'):
+    >>> x = np.linspace(-2, 2, num=20)
+    >>> y = x
+    >>> y_int = integrate.cumulative_simpson(y, x, initial=0)
+    >>> plt.plot(x, y_int, 'ro', x, y[0] + 0.5 * x**2, 'b-')
```
Please use OOP API from Matplotlib.

nprav replied:
Addressed in [fa4a01d](https://github.com/scipy/scipy/pull/18151/commits/fa4a01d850c03d13d114ca202b02faf986611899)

tupui commented on scipy/integrate/_quadrature.py:
```diff
@@ -662,6 +664,254 @@ def simpson(y, x=None, dx=1.0, axis=-1, even='avg'):
+
+    References
+    ----------
+    .. [1] Wikipedia page: https://en.wikipedia.org/wiki/Simpson's_rule
```
Can we have a reference from an article (additionally at least.)

nprav replied:
Added in [92fb1ef](https://github.com/scipy/scipy/pull/18151/commits/92fb1efb267ff8b402c9f693105051bdb9f8ff85)

tupui commented on scipy/integrate/_quadrature.py:
```diff
@@ -662,6 +664,254 @@ def simpson(y, x=None, dx=1.0, axis=-1, even='avg'):
+    then the result is exact only if the function is a polynomial of order 2
+    or less.
+
+    The composite Simpson's 1/3 method is outlined in [1]. A quadratic relationship
```
We might want to add the mathematical formulation here.
```suggestion
    The composite Simpson's 1/3 method is outlined in [1]_. A quadratic relationship
```

nprav replied:
Added in [92fb1ef](https://github.com/scipy/scipy/pull/18151/commits/92fb1efb267ff8b402c9f693105051bdb9f8ff85)

tupui commented on scipy/integrate/_quadrature.py:
```diff
@@ -662,6 +664,254 @@ def simpson(y, x=None, dx=1.0, axis=-1, even='avg'):
+def cumulative_simpson(y, x=None, dx=1.0, axis=-1, initial=None):
+    """
+    Cumulatively integrate y(x) using the composite Simpson's 1/3 rule.
+    If x is None, spacing of dx is assumed.
```
Not the place to describe this.
```suggestion
```

nprav replied:
Removed in [be6e1d7](https://github.com/scipy/scipy/pull/18151/commits/be6e1d75e7863d46c297ad24a9929392873dcf3a). The required detail is already in the kwarg description section of the docstring.

tupui commented on scipy/integrate/tests/test_quadrature.py:
```diff
@@ -397,3 +397,240 @@ def func(x):
+            y_shape[axis] = slice(1, None)
+            int_y = int_y[tuple(y_shape)]
+
+        with pytest.raises(ValueError):
```
Use a match statement to be sure the correct exception is raised.

nprav replied:
Addressed in [be6e1d7](https://github.com/scipy/scipy/pull/18151/commits/be6e1d75e7863d46c297ad24a9929392873dcf3a)

tupui replied:
I am a bit confused, the match don't correspond to the error messages.

nprav replied:
Did not fully understand you. Could you explain the issue with an example?

I added a few more unit tests to cover all the ValueErrors that are raised by the function.

tupui replied:
What I mean is that, your function has raise statements. In the tests, we should be able to see all of these statements being verified (we only want to catch the new ones, not things coming from elsewhere.) Unless I am mistaking, I could not find some of the new exception messages in the tests.

nprav replied:
Got it, you were correct, not all the raise statements were covered. 
I have added more tests to address all the statements in [e5de535](https://github.com/scipy/scipy/pull/18151/commits/e5de5352353ba28a62044cd6708b00ca26b7c002)

tupui replied:
Thank you, great 👍 Now I would only add tests for the new exceptions. So if you have 5 new exceptions, there should just be 5 tests.

nprav replied:
I have added 8 exception tests for 6 new exceptions. 2 additional unit tests are required to ensure one of the raises is working for different types of input parameters (1D vs multi-D inputs and non-default integration axis).

tupui commented on scipy/integrate/_quadrature.py:
```diff
@@ -662,6 +664,254 @@ def simpson(y, x=None, dx=1.0, axis=-1, even='avg'):
+    >>> plt.plot(x, y_int, 'ro', x, y[0] + 0.5 * x**2, 'b-')
+    >>> plt.show()
+
+    """
+
+    y = np.asarray(y)
```
```suggestion
    """
    y = np.asarray(y)
```

nprav replied:
Addressed in [be6e1d7](https://github.com/scipy/scipy/pull/18151/commits/be6e1d75e7863d46c297ad24a9929392873dcf3a)

tupui reviewed (changes requested):
Thank you for the update @nprav.

This PR would need a review from a maintainer with expertise in interpolate. I personally don't have the bandwidth at the moment to help I am afraid. Still I left a few comments to help out a bit. I will let the CI run so that we also have an idea.

I don't think this will make it in the next release unless a maintainer with the relevant expertise push for it. We have a lot of PR/issues tagged for 1.11 already and we just have 2 more weeks before the branch freeze.

tupui reviewed (commented):
Thank you for the update @nprav

I know our linter is not very helpful at the moment with these, but all lines should be less than 80 chars (you could use something like Black or an IDE which would do this for you.)

For the rest, as I said I am afraid I don't really have time. I added a milestone so that it hopefully gets a bit more attention.

nprav commented:
Hi @tupui , is there anything else I can do to help get this PR merged?

tupui commented:
Hi @nprav as I said above, I am afraid I am not the right person to ping. And right now I really don't have much time.

Still, besides the conflict to solve it looks like it's in a good shape. The milestone is there and should help to get more attention as we approach the release phase (still a few months to go.)

lucascolley commented:
@steppi perhaps you could give this a look? Or should we bump the milestone?

steppi commented:
> @steppi perhaps you could give this a look? Or should we bump the milestone?

We branch December 6th? Yeah, I think I can get to it on time.

nprav commented on scipy/integrate/tests/test_quadrature.py:
```diff
@@ -524,6 +524,284 @@ def func(x):
+            y_shape[axis] = slice(1, None)
+            int_y = int_y[tuple(y_shape)]
+
+        assert_almost_equal(
```
Thanks Lucas, updated these

mdhaber commented:
Thanks @nprav, @lucascolley. I've chatted with Albert about reviewing this one for him, since I have a few PRs that could use his attention. 

Please let CI finish so we can see what happens on CircleCI. It looks like the docs have not build successfully yet, so that is one of the first things to address.

I'd also like to see a unit test of `cumulative_simpson` against `simpson` - I'd expect `cumulative_simpson` to return something like:
```python3
import numpy as np
from scipy import integrate

x = np.linspace(0, 1)
y = np.sin(x)
Y = [integrate.simpson(y[:i], x[:i]) for i in range(1, len(y)+1)]
```
If you have some time and want to get started, could you draft that test? (But please don't push yet.)

lucascolley commented on scipy/integrate/_quadrature.py:
```diff
@@ -822,6 +824,294 @@ def simpson(y, *, x=None, dx=1.0, axis=-1, even=_NoValue):
+        The coordinate to integrate along. Must be monotonically increasing.
+        If None (default), use spacing `dx` between consecutive elements in 
+        `y`.
+    dx : scalar | array_like, optional
```
```suggestion
    dx : scalar or array_like, optional
```

lucascolley commented on scipy/integrate/_quadrature.py:
```diff
@@ -822,6 +824,294 @@ def simpson(y, *, x=None, dx=1.0, axis=-1, even=_NoValue):
+        `axis`.
+    axis : int, optional
+        Specifies the axis to integrate along. Default is -1 (last axis).
+    initial : scalar | array_like, optional
```
```suggestion
    initial : scalar or array_like, optional
```

lucascolley reviewed (commented):
Docs have built fine, feel free to go ahead with pushing in changes to the tests.

I think that the convention is to use `or` rather than `|` in docstrings for types of arguments.

mdhaber commented on scipy/integrate/_quadrature.py:
```diff
@@ -822,6 +824,294 @@ def simpson(y, *, x=None, dx=1.0, axis=-1, even=_NoValue):
+
+    Note that the integral formulae in [1]_ and [2]_ use different variables.
+
+    For samples that are equally spaced the result is exact if the function
```
```suggestion
    For samples that are equally spaced, the result is exact if the function
```

mdhaber commented on scipy/integrate/_quadrature.py:
```diff
@@ -822,6 +824,294 @@ def simpson(y, *, x=None, dx=1.0, axis=-1, even=_NoValue):
+        )
+
+    if x is not None:
+        x = np.asarray(x)
+        if x.ndim == 1 and y.shape[axis] == x.shape[0]:
+            x_shape = [1] * y.ndim
+            x_shape[axis] = -1
+            x = x.reshape(tuple(x_shape))
+        elif x.shape != y.shape:
+            raise ValueError(
+                "If given, shape of `x` must be the same as `y` or 1-D with "
+                "the same length as `y` along `axis`."
+            )
```
Please use [`numpy.broadcast_arrays`](https://numpy.org/doc/stable/reference/generated/numpy.broadcast_arrays.html) and accept any arrays that are mutually broadcastable. Do this near the top of the function. It takes care of converting array-like input into arrays, too.

*Update: Never mind... I guess `simpson` and `cumulative_trapezoid` don't follow normal broadcasting rules. I will make a different suggestion later.*

mdhaber commented on scipy/integrate/_quadrature.py:
```diff
@@ -822,6 +824,294 @@ def simpson(y, *, x=None, dx=1.0, axis=-1, even=_NoValue):
+    y = np.asarray(y)
+
+    # validate y and axis
+    if axis < -1 or axis >= y.ndim:
```
I think this is too restrictive. After broadcasting the arrays to the same shape, consider:
```suggestion
    if not (-y.ndim <= axis < y.ndim):
```

Opinion, but you don't need to change it: better yet, let NumPy take care of this for us. I find that it's always easier to read and write code involving N-D arrays with an assumption about `axis`, so I always move `axis` to the end:
```python3
try:
    x = np.moveaxis(x, axis, -1)
    y = np.moveaxis(y, axis, -1)
    axis = -1
except numpy.exceptions.AxisError as e:
    raise ValueError(message) from e
```
 If NumPy complains, this re-raises the error with something SciPy-specific. Otherwise, you're good to go.

nprav replied:
Applied the axis change to the implementation functions.

mdhaber replied:
The change to using axis to `-1` looks great. This comment is still not completely resolved because some indices < -1 are valid, but this would still raise an error. Why not change all the axes here as suggested? Actually, if you do that, I think then you really *could* just use `broadcast_arrays` to take care of all the shape compatibility stuff after all.

mdhaber commented on scipy/integrate/_quadrature.py:
```diff
@@ -822,6 +824,294 @@ def simpson(y, *, x=None, dx=1.0, axis=-1, even=_NoValue):
+        res = _cumulative_simpson_unequal_intervals(y, x, axis=axis)
+
+    else:
+        dx = np.asarray(dx)
+        final_dx_shape = tupleset(y.shape, axis, y.shape[axis] - 1)
+        alt_input_dx_shape = tupleset(y.shape, axis, 1)
+
+        if dx.ndim == 0:
+            dx = np.ones(final_dx_shape) * dx
+        elif dx.shape == alt_input_dx_shape:
+            dx = np.repeat(dx, y.shape[axis] - 1, axis=axis)
+        else:
+            raise ValueError(
+                "`dx` must either be numeric or have the same shape as `y` "
+                "but with only 1 point along `axis`."
+            )
+        res = _cumulative_simpson_equal_intervals(y, dx, axis=axis)
+
+    if initial is not None:
+        initial = np.asarray(initial)
+        alt_initial_input_shape = tupleset(y.shape, axis, 1)
+        if initial.ndim == 0:
+            initial = np.ones(y.shape) * initial
+        elif initial.shape == alt_initial_input_shape:
+            initial = np.repeat(initial, y.shape[axis], axis=axis)
+        else:
+            raise ValueError(
+                "`initial` must either be numeric or have the same shape as "
+                "`y` but with only 1 point along `axis`."
+            )
+
+        slice1 = tupleset((slice(None),) * y.ndim, axis, slice(1, None))
+        initial[slice1] = initial[slice1] + res
+        res = initial
```
I think we can simplify this a lot with the help of `np.broadcast_arrays`. To make sure I understand what you're after, is this exactly true?
> `dx` must either be numeric or have the same shape as `y` but with only 1 point along `axis`.

To be clear, suppose `x` and `y` are valid inputs. Then you would not allow the user to use:
```
dx = np.diff(x)
```
as `dx` instead of passing `x`, right?

nprav replied:
Your understanding is correct regarding dx. It is consistent with cumulative_trapezoid.

I agree with you though, the np.broadcast* functions should be used to simplify this. Will get back to you after this updates.

mdhaber replied:
Well... it would have really simplified things if the predecessors followed normal broadcasting rules, but it doesn't. So I'd rather just make this one consistent with the other three. I'd suggest factoring out the array processing code from `cumulative_trapezoid` into a function and using it in both functions.

mdhaber reviewed (commented):
A few things to start with.

mdhaber commented on scipy/integrate/_quadrature.py:
```diff
@@ -822,6 +824,294 @@ def simpson(y, *, x=None, dx=1.0, axis=-1, even=_NoValue):
+
+    # Calculate integral over the h1 subintervals.
+    # Calculate for all but last subinterval of y.
+    sub_integrals_h1 = dx[:-1] / 3.0 * \
+        (5.0 * y[:-2] / 4.0 + 2.0 * y[1:-1] - y[2:] / 4.0)
```
These are easy to fit on one line by omitting the `0`s. TBH, I think you can omit the `.0`s.

nprav replied:
Updated these to use parentheses.
The .0s ensure conversion to floats. There is an edge case where if the inputs are int arrays, the integration output may end up as an int array as well giving inaccurate results. Having the .0s ensures the result returns an accurate result with a float array.

mdhaber replied:
You can still omit the `0`s. 

I'd suggest ensuring float dtype from the outset. It avoids such edge cases, and it doesn't require repeated promotion of the whole array to float before each operation.

```
import numpy as np
x = np.arange(999999)
%timeit x*5.
# 3.13 ms ± 346 µs per loop
x = x.astype(float)
%timeit x*5
# 1.43 ms ± 443 µs per loop
```

nprav replied:
I have removed the .0s and fit the lines you highlighted to a single line.
Regarding casting to float, earlier @MatteoRaso raised a valid concern; what if the input is float128? If we cast astype(float), the array will be converted to float64 which is lower precision than the input (https://github.com/scipy/scipy/pull/18151#discussion_r1167135755).

To address this, I am now only casting to float if the input is an int array (https://github.com/scipy/scipy/pull/18151/files#diff-2ff88b8913ad74b9ea3fe1fe02d6d49164855189168b767076af2e94206c3a2cR938).

mdhaber replied:
The purpose of my code was to demonstrate the advantage of casting to a floating point type. I didn't mean to cast to `float` unconditionally. Using `issubdtype` to detect integers is exactly what I would have recommended, so this looks good.

mdhaber commented on scipy/integrate/_quadrature.py:
```diff
@@ -822,6 +824,294 @@ def simpson(y, *, x=None, dx=1.0, axis=-1, even=_NoValue):
+        if dx.ndim == 0:
+            dx = np.ones(final_dx_shape) * dx
+        elif dx.shape == alt_input_dx_shape:
+            dx = np.repeat(dx, y.shape[axis] - 1, axis=axis)
```
`repeat` copies data. If you don't need to modify the array, prefer `np.broadcast_to`.

nprav commented on scipy/integrate/_quadrature.py:
```diff
@@ -822,6 +824,294 @@ def simpson(y, *, x=None, dx=1.0, axis=-1, even=_NoValue):
+    tplquad : triple integrals
+    romb : integrators for sampled data
+    ode : ODE integrators
+    odeint : ODE integrators
```
Applied your suggested change.

nprav commented:
> I'd also like to see a unit test of `cumulative_simpson` against `simpson`. By analogy with `trapezoid`/`cumulative_trapezoid`, I'd expect a test something like:
> 
> ```python
> import numpy as np
> from scipy import integrate
> 
> def cumulative_simpson_reference(y, x):
>     return np.asarray([integrate.simpson(y[:i], x=x[:i]) for i in range(2, len(y)+1)])
> 
> x = np.linspace(0, 1, 10)
> y = np.sin(x)
> 
> res = integrate.cumulative_simpson(y, x=x)
> ref = cumulative_simpson_reference(y, x)
> 
> np.testing.assert_allclose(res, ref, rtol=5e-16)
>
> abs(res - ref) > 1e-6
> # array([ True, False,  True, False,  True, False,  True, False, False])
> ```



The difference between simpson and cumulative simpson that you observed is expected. 
The main reason for the difference is that you are passing the entire array (y) to cumulative_simpson, whereas simpson is only getting a slice of the array at each point (i). 

When i = 2 (index 0 of abs(res - ref)), `simpson` only receives 2 points. Hence it uses the the trapezoidal rule. This causes the discrepancy.

When i = 4, 6, 8, there are an odd number of intervals in y. `simpson` calculates the last interval using a different formula [(see here)](https://github.com/nprav/scipy/blob/cumulative-simpson/scipy/integrate/_quadrature.py#L758-L795).

It is better to compare `simpson` and `cumulative_simpson` by passing identical inputs:

```python
import numpy as np
from scipy import integrate

def simpson_reference(y, x):
    return np.asarray([integrate.simpson(y[:i], x=x[:i]) for i in range(3, len(y)+1)])

def cumulative_simpson_reference(y, x):
    return np.asarray([integrate.cumulative_simpson(y[:i], x=x[:i])[-1] for i in range(3, len(y)+1)])

x = np.linspace(0, 1, 10)
y = np.sin(x)

res = simpson_reference(y, x)
ref = cumulative_simpson_reference(y, x)

np.testing.assert_allclose(res, ref, rtol=5e-16)

abs(res - ref)
# array([3.46944695e-18, 6.93889390e-18, 0.00000000e+00, 2.77555756e-17,
#      2.77555756e-17, 5.55111512e-17, 5.55111512e-17, 5.55111512e-17])
```

This passes as expected.

mdhaber commented:
I vaguely remembered there being a difference between odd and even numbers of points. So that leads to:
- Are there any reasonable choices that could have been made in the code here, or changes that should be made in `simpson`, that would resolve the discrepancy? 
- Is one (presumably `cumulative_trapezoid`, which has access to more information) more accurate?
@andyfaff Is this discrepancy between `cumulative_simpson` and iteratively calling `simpson` OK?

nprav commented:
> I vaguely remembered there being a difference between odd and even numbers of points. Good. But that leads to:
> 
>     * Are there any reasonable choices that could have been made in the code here, or changes that should be made in `simpson`, that would resolve the discrepancy?
> 

Actually, Simpson was recently improved to make it more accurate with an even number of points (https://github.com/scipy/scipy/pull/18209). Nothing further can be done for this discrepancy; it is theoretically expected.

>     * Is one (presumably `cumulative_trapezoid`, which has access to more information) more accurate?

For data sampled from continuously differentiable functions, simpson is much more accurate than cumulative_trapezoid ([error comparisons at the end here](https://github.com/nprav/misc/blob/main/cumulative_simpson/Cumulative%20Simpson%20algorithm%20derivation.ipynb)). If we want to assert against outputs from cumulative_trapezoid, we would need to experiment with reducing the tolerances.

>     * Even if they can't/shouldn't be forced to agree, I still think it is valuable for testing to have a reference for the complete output - not just the last element - of `cumulative_simpson` in terms of `simpson`. Can my suggestion be modified to always use `simpson` with an even number of intervals, and then the last step would be taken with a separate (presumably simple) formula?

Added two property-based tests comparing simpson and cumulative_simpson. The comparisons are only done at the even indices and the last index. Theoretically these indices will match. Let me know if this is what you were looking for.

mdhaber commented:
Ok. We should document the discrepancy to avoid bug reports - we'll just note that intermediate results of the cumulative version are more accurate (assuming a well behaved underlying function) than calling `simpson` with the same upper limit.

For the reference implementation in the tests, I suppose we can just compare for inputs with odd length.

mdhaber commented:
@nprav I submitted nprav/scipy#1 for your review. Please take a look. Also, can you add comments in GH about the purpose of each test? Usually there is a comment about what the test as a whole is for, but sometimes I can't tell what the differences between the parametrized cases are supposed to do.

nprav commented on scipy/integrate/tests/test_quadrature.py:
```diff
@@ -524,6 +527,399 @@ def func(x):
+        ("y_func", "x", "dx", "initial", "int_y_func"),
+        [
+            (lambda y: y, None, 1.0, 0, lambda y: y**2 / 2),
+            (lambda y: y, None, 0.5, 5, lambda y: y**2 / 2 + 5),
```
Test simpson 1d with dx (_cumulative_simpson_equal_intervals) with initial

nprav commented on scipy/integrate/tests/test_quadrature.py:
```diff
@@ -524,6 +527,399 @@ def func(x):
+                1.0,
+                None,
+                lambda y: y**2 / 2,
+            ),
```
Test simpson 1d with x(_cumulative_simpson_unequal_intervals)

nprav commented on scipy/integrate/tests/test_quadrature.py:
```diff
@@ -524,6 +527,399 @@ def func(x):
+                None,
+                lambda y: y**2 / 2,
+            ),
+            (lambda y: y**2, None, 0, None, lambda y: 0 * y),
```
Test edge case, dx = 0

nprav commented on scipy/integrate/tests/test_quadrature.py:
```diff
@@ -524,6 +527,399 @@ def func(x):
+                None,
+                -1,
+                lambda y: y**3 / 3 + 1 / 3,
+            ),
```
cumulative_simpson, Multi dimensional inputs with x.

nprav commented on scipy/integrate/tests/test_quadrature.py:
```diff
@@ -524,6 +527,399 @@ def func(x):
+                1 / 3,
+                -1,
+                lambda y: y**3 / 3,
+            ),
```
cumulative_simpson, Multi dimensional input for x with initial (ndim = 1)

nprav commented on scipy/integrate/tests/test_quadrature.py:
```diff
@@ -524,6 +527,399 @@ def func(x):
+                np.array([[1 / 3], [1 / 3], [1 / 3]]),
+                -1,
+                lambda y: y**3 / 3,
+            ),
```
cumulative_simpson, Multi dimensional input for x with initial (ndim > 1)

nprav commented on scipy/integrate/tests/test_quadrature.py:
```diff
@@ -524,6 +527,399 @@ def func(x):
+        x = np.array([input_x]*3)
+        y = x**2
+        int_y = x**3 / 3
+        assert_allclose(
```
Multidimensional inputs with x.ndim < y.ndim

nprav commented:
@mdhaber 

1. Updated the tests comparing simpson with cumulative_simpson by calculating the expected difference at the even intervals. So now we are comparing every index except for the first (simpson uses trapezoidal rule here).
2. Added GH comments describing test cases.
3. Committed your PR + exceptions test update

nprav commented:
> Thanks @nprav. I think we'll finish this before we branch. I just want to make sure it's all expressed in a way that can easily be maintained. It looks like you've been working on this for a long time and have gone through several reviewers, so I don't want to push your patience. If there are other things (like the input validation and corresponding tests) that are technically correct but can be expressed more clearly (it seems that you agreed that the changes were improvements), are you willing to help with that?

Yup, I'm willing to help with any specific comments.

mdhaber commented on scipy/integrate/_quadrature.py:
```diff
@@ -822,6 +825,305 @@ def simpson(y, *, x=None, dx=1.0, axis=-1, even=_NoValue):
+    h2 subintervals. Returns the cumulative sum using
+    composite Simpson's rule. Assumes the axis of summation is -1.
+    """
+    # Set integral of odd h1 intervals = 0
+    sub_integrals_h1[..., 1::2] = 0.0
+    # Set integral of odd h2 intervals = 0, except for the last interval
+    sub_integrals_h2[..., 1:-1:2] = 0.0
+    sub_integrals = np.concatenate(
+        [
+            # Integral over first subinterval can only be calculated from 
+            # formula for h1
+            sub_integrals_h1[..., 0:1],
+            # Combine sub_integral arrays (at every index, i, either
+            # sub_integrals_h1[i] = 0 or sub_integrals_h2[i] = 0)
+            (sub_integrals_h1[..., 1:] + sub_integrals_h2[..., :-1]),
+            # Integral over last subinterval can only be calculated from 
+            # formula for h2
+            sub_integrals_h2[..., -1:],
+        ],
+        axis=-1,
+    )
```
Isn't this all just like:
```python3
sub_integrals = np.append(sub_integrals_h1, sub_integrals_h2[-1:], axis=-1)
sub_integrals[1::2] = sub_integrals_h2[::2]
```
?

Also, it looks like we're throwing away half the values that were calculated. Why don't we just calculate the ones that are needed?

nprav replied:
Optimized this further. I used np.concatenate instead of np.append as I found it was more efficient.

> it looks like we're throwing away half the values that were calculated. Why don't we just calculate the ones that are needed?

I have refactored the internal functions to only integrate the intervals required. 
It is slightly more efficient than if we just calculate the entire array and discard half (see optimized version 1 vs version 2 at the end of [this gist](https://gist.github.com/nprav/1c8a82738cb4f8ceffe740a5500b8d61)). 
But, it does make the internal functions and function signatures a bit more complicated.
Let me know if you can think of a better way to implement this.

mdhaber replied:
Let's do optimized version 1. Simplicity is what's most important now, if we're trying to get this merged before branching 1.13.

(I considered starting with an array of zero of the right shape then populating it, as you have. I ended up with what I wrote because it had the same complexity and was simpler, which was the motivation. But what you have in optimized verison 1 is fine. You can use [`np.empty`](https://numpy.org/doc/stable/reference/generated/numpy.empty.html) instead of `np.zeros`, even though it rarely make a practical difference.)

mdhaber reviewed (commented):
Here are some thoughts on the actual implementation to make the code easier to verify and improve performance.

nprav commented on scipy/integrate/_quadrature.py:
```diff
@@ -822,6 +825,305 @@ def simpson(y, *, x=None, dx=1.0, axis=-1, even=_NoValue):
+        -y1 * (h2**3 / (6 * h1)) / (h1 + h2)
+        + y2 * (3 * h1 + h2) * h2 / (6 * h1)
+        + y3 * (3 * h1 + 2 * h2) / (h1 + h2) * h2 / 6
+    )
```
Adopted these in latest commit

mdhaber commented:
```
FAILED scipy/integrate/tests/test_quadrature.py::TestCumulativeSimpson::test_simpson_exceptions[`axis=4` is not valid for `y` with `y.ndim=1`-kwarg_update4] - AttributeError: module 'numpy' has no attribute 'AxisError'
```
Not sure why this is failing on just one platform, but it sounds like `AxisError` subclasses `ValueError` and `IndexError`. Maybe except `IndexError` if we can't use `AxisError`.

mdhaber commented on scipy/integrate/_quadrature.py:
```diff
@@ -822,6 +825,271 @@ def simpson(y, *, x=None, dx=1.0, axis=-1, even=_NoValue):
+    integrals for the subintervals are derived (see [2]_ for derivation 
+    method):
+
+    .. math::
+        \int_{x_{1}}^{x_{2}} y(x) dx\ = \int_{0}^{h_{1}} y(x) dx\
+            = \frac{h_{1}(2h_{1}+3h_{2})}{6(h_{1}+h_{2})} y_{1}
+            + \frac{h_{1}(h_{1}+3h_{2})}{6h_{2}} y_{2}
+            - \frac{h_{1}^{3}}{6h_{2}(h_{1}+h_{2})} y_{3}
+
+    .. math::
+        \int_{x_{2}}^{x_{3}} y(x) dx\ = \int_{h_{1}}^{h_{1}+h_{2}} y(x) dx\
+            = - \frac{h_{2}^{3}}{6h_{1}(h_{1}+h_{2})} y_{1}
+            + \frac{h_{2}(3h_{1}+h_{2})}{6h_{1}} y_{2}
+            + \frac{h_{2}(3h_{1}+2h_{2})}{6(h_{1}+h_{2})} y_{3}
+
+    .. math::
+        \int_{x_{1}}^{x_{3}} y(x) dx\ = \int_{0}^{h_{1}+h_{2}} y(x) dx\
+            = \frac{(h_{1}+h_{2})(2h_{1}-h_{2})}{6h_{1}} y_{1}
+            + \frac{(h_{1}+h_{2})^{3}}{6h_{1}h_{2}} y_{2}
+            - \frac{(h_{1}-2h_{2})(h_{1}+h_{2})}{6h_{2}} y_{3}
+
+    Note that the integral formulae in [1]_ and [2]_ use different variables.
```
Let's replace all these with just Equation 8 from [2]. We can explain how the other integrals (from x2 to x3 and from x1 to x3) are calculated from that.

nprav replied:
Updated these equations to put them in the form of [2]. Just saw your comment. Since we are using the \int_{x_{2}}^{x_{3}} y(x) dx equation, I think it should be included.

nprav commented on scipy/integrate/tests/test_quadrature.py:
```diff
@@ -524,6 +527,324 @@ def func(x):
+        with pytest.raises(ValueError, match=message):
+            cumulative_simpson(**dict(kwargs0, **kwarg_update))
+
+    def _get_theoretical_diff_between_simps_and_cum_simps(self, y, x):
```
@mdhaber , 
I want to update this function by importing and using integrate._cumulative_simpson_unequal_intervals instead of adding the maths here.
But the function is not in the __all__ list of _quadrature.py so it does not get imported. It also should not be since it should not be accessible by users.
I can get around this by moving _cumulative_simpson_unequal_intervals  to a separate file.
Is there another way around this?

lucascolley replied:
You can add `from scipy.integrate._quadrature import _cumulatively_sum_simpson_integrals` at the top of the file.

mdhaber commented on scipy/integrate/_quadrature.py:
```diff
@@ -822,6 +825,270 @@ def simpson(y, *, x=None, dx=1.0, axis=-1, even=_NoValue):
+    where the widths of the 2 subintervals are:
+    :math:`h_1 = x_2 - x_1` and :math:`h_2 = x_3 - x_2`.
+
+    Assuming a quadratic relationship over the 3 points, the following 
+    integrals for the subintervals are derived (formulae (6) and (8) of [2]_):
+
+    .. math::
+        \int_{x_1}^{x_3} y(x) dx\ = \frac{x_3-x_1}{6}\left[\
+        \left\{2-\frac{x_3-x_2}{x_2-x_1}\right\} y_1 + \
+        \frac{(x_3-x_1)^2}{(x_2-x_1)(x_3-x_2)} y_2 + \
+        \left\{2-\frac{x_2-x_1}{x_3-x_2}\right\} y_3\right]
+
+    .. math::
+        \int_{x_1}^{x_2} y(x) dx\ = \frac{x_2-x_1}{6}\left[\
+        \left\{3-\frac{x_2-x_1}{x_3-x_1}\right\} y_1 + \
+        \left\{3 + \frac{(x_2-x_1)^2}{(x_3-x_2)(x_3-x_1)} + \
+        \frac{x_2-x_1}{x_3-x_1}\right\} y_2 - \
+        \frac{(x_2-x_1)^2}{(x_3-x_2)(x_3-x_1)} y_3\right]
+
+    .. math::
+        \int_{x_2}^{x_3} y(x) dx\ = \frac{x_3-x_2}{6}\left[\
+        - \frac{(x_3-x_2)^2}{(x_2-x_1)(x_3-x_1)} y_1 + \
+        \left\{3 + \frac{(x_3-x_2)^2}{(x_2-x_1)(x_3-x_1)} + \
+        \frac{x_3-x_2}{x_3-x_1}\right\} y_2 + \
+        \left\{3-\frac{x_3-x_2}{x_3-x_1}\right\} y_3\right]
+
+    Note that the integral formulae in [1]_ and [2]_ use different variables.
+
+    For samples that are equally spaced, the result is exact if the function
+    is a polynomial of order 3 or less [1]_. If the samples are not equally 
+    spaced, then the result is exact only if the function is a polynomial of 
+    order 2 or less.
```
Doesn't this more closely match what we are actually doing, though?

```suggestion
    Assuming a quadratic relationship over the 3 points, the integral over 
    the subinterval between :math:`x_1` and :math:`x_2` is given by formula
    (8) of [2]_:
    
    .. math::
        \int_{x_1}^{x_2} y(x) dx\ = \frac{x_2-x_1}{6}\left[\
        \left\{3-\frac{x_2-x_1}{x_3-x_1}\right\} y_1 + \
        \left\{3 + \frac{(x_2-x_1)^2}{(x_3-x_2)(x_3-x_1)} + \
        \frac{x_2-x_1}{x_3-x_1}\right\} y_2 - \
        \frac{(x_2-x_1)^2}{(x_3-x_2)(x_3-x_1)} y_3\right]

    The integral between :math:`x_2` and :math:`x_3` is given by swapping
    appearances of :math:`x_1` and :math:`x_3`.
    
    For samples that are equally spaced, the result is exact if the function
    is a polynomial of order 3 or less [1]_ and the number of subintervals
    is even. Otherwise, the integral is exact for polynomials of  order 2 or
    less.
```

I don't think we need to mention the integral between `x1` and `x3` because you already said that the overall integral is calculated by cumulative summation over the sub-intervals.

mdhaber commented on scipy/integrate/_quadrature.py:
```diff
@@ -822,6 +825,270 @@ def simpson(y, *, x=None, dx=1.0, axis=-1, even=_NoValue):
+    The composite Simpson's 1/3 method can be used to approximate the definite 
+    integral of a sampled input function :math:`y(x)` [1]_. The method assumes 
+    a quadratic relationship over the interval containing any 3 consecutive 
+    sampled points. The integral is estimated separately for each interval
```
```suggestion
    sampled points. The integral is estimated separately for each subinterval
```

mdhaber commented on scipy/integrate/_quadrature.py:
```diff
@@ -822,6 +825,270 @@ def simpson(y, *, x=None, dx=1.0, axis=-1, even=_NoValue):
+    The composite Simpson's 1/3 method can be used to approximate the definite 
+    integral of a sampled input function :math:`y(x)` [1]_. The method assumes 
+    a quadratic relationship over the interval containing any 3 consecutive 
+    sampled points. The integral is estimated separately for each interval
+    and then cumulatively summed to obtain the final result.
```
```suggestion
    sampled points.
```
moving this below

mdhaber commented:
After https://github.com/scipy/scipy/pull/18151#pullrequestreview-1766370977 and nprav/scipy#2 are considered, I think this is good to go.

mdhaber reviewed (approved):
There are some details to improve, but let's get this in. Thanks @nprav!

nprav commented:
Thank you all for helping to add this to SciPy! @mdhaber @tupui @andyfaff @lucascolley @lorentzenchr @MatteoRaso @j-bowhay

lucascolley commented:
thanks @nprav and congratulations on your first contribution to SciPy! Keep them coming :)

mdhaber commented on scipy/integrate/_quadrature.py:
```diff
@@ -822,6 +825,260 @@ def simpson(y, *, x=None, dx=1.0, axis=-1, even=_NoValue):
+    the subinterval between :math:`x_1` and :math:`x_2` is given by formula
+    (8) of [2]_:
+    
+    .. math::
+        \int_{x_1}^{x_2} y(x) dx\ = \frac{x_2-x_1}{6}\left[\
+        \left\{3-\frac{x_2-x_1}{x_3-x_1}\right\} y_1 + \
+        \left\{3 + \frac{(x_2-x_1)^2}{(x_3-x_2)(x_3-x_1)} + \
+        \frac{x_2-x_1}{x_3-x_1}\right\} y_2 - \
+        \frac{(x_2-x_1)^2}{(x_3-x_2)(x_3-x_1)} y_3\right]
```
I think we need to line wrap

<img width="955" alt="image" src="https://github.com/scipy/scipy/assets/6570539/c4dacb81-4289-4a01-8c3f-3e741f4a9ef9">

nprav replied:
Added line wrap, but could not get it specifically aligned. Do you have an example of this elsewhere in the docs?
![image](https://github.com/scipy/scipy/assets/7060774/e0f15e6c-963f-4dc6-bc8b-10f16f48aa27)

mdhaber replied:
I'm not sure. This is fine with me. If someone else wants to fight with Sphinx, that's fine with me, but I'm happy as long as it doesn't extend past the right edge.

mdhaber commented on scipy/integrate/_quadrature.py:
```diff
@@ -822,6 +825,260 @@ def simpson(y, *, x=None, dx=1.0, axis=-1, even=_NoValue):
+        message = f"`axis={axis}` is not valid for `y` with `y.ndim={y.ndim}`."
+        raise ValueError(message) from e
+
+    if y.shape[-1] < 3:
+        raise ValueError(
+            "At least 3 points are required along the axis of integration "
+            "to use the composite Simpson's method."
+        )
```
This is not an error in `simpson` 
```python3
integrate.simpson([1], x=[1])
# 0.0
```

or the analogous case in `cumulative_trapezoid`:
```python3
from scipy import integrate
integrate.cumulative_trapezoid([1], x=[1], initial=0)
# array([0.])
```

So I think we should make this work when `y.shape[-1] > 0`, and raise the error here when `y.shape[-1] = 0`. Because of the function's behavior of reducing the length by one, I don't think there is a correct result to be returned when `y.shape[-1] = 0` and `initial` is not specified. With `initial` specified, it could be argued that an empty array would be the correct result, but I think this would be confusing. (If only these could follow normal broadcasting rules and always return an output with the same shape as `y` : / It would all be so much simpler!)

nprav replied:
To address this, shall we do the following then:
0< y.shape[-1] < 3 ---> return result from cumulative_trapezoid
y.shape[-1] --> raise error

This would match the behaviour from simpson. Simpson also just returns trapezoid result when y.shape[-1] = 2

mdhaber replied:
Yeah, that's a good idea. Actually, just let the function this calls raise the error for 0 length (and make sure it's understandable in this context, too).

nprav replied:
Addressed in #19709
cumulative_trapezoid does not have an explicit error for when y.shape[-1] <=0. Hence added one in in cumulative_simpson: https://github.com/scipy/scipy/pull/19709/files#diff-2ff88b8913ad74b9ea3fe1fe02d6d49164855189168b767076af2e94206c3a2cR1042

mdhaber replied:
Yeah, so then why not add the error to `cumulative_trapezoid`, since it's desirable there, too?

mdhaber commented on scipy/integrate/_quadrature.py:
```diff
@@ -822,6 +825,260 @@ def simpson(y, *, x=None, dx=1.0, axis=-1, even=_NoValue):
+
+        x = np.broadcast_to(x, y.shape) if x.ndim == 1 else np.swapaxes(x, axis, -1)
+        dx = np.diff(x, axis=-1)
+        if np.any(dx <= 0):
+            raise ValueError("Input x must be monotonically increasing.")
```
The requirement should really be non-decreasing.

```python3
from scipy import integrate
integrate.simpson([1, 2], x=[1, 1])
# 0
```

nprav replied:
Monotonically increasing = non-decreasing.
Can change to non-decreasing for clarity

nprav replied:
Updated to 'strictly increasing' in #19709 due to the issue raised in https://github.com/scipy/scipy/pull/18151#issuecomment-1859087314

mdhaber commented on scipy/integrate/tests/test_quadrature.py:
```diff
@@ -524,6 +528,226 @@ def func(x):
+        res = cumulative_simpson(y, dx=0)
+        assert_equal(res, 0)
+
+        # Should add tests of:
+        # - all elements of `x` identical
+        # - y of size 1 and 2
+        # These should work as they do for `simpson`
```
Tests corresponding with the behavior changes above would go here.

nprav replied:
Added tests for y of size 1 and 2 in #19709: https://github.com/scipy/scipy/pull/19709/files#diff-59a2aea1061e8fc7c39d851f35d26148991fc0d48917342b58af96c078254093R664-R699

mdhaber reviewed (commented):
@nprav Here are some things to consider for follow-up work. If you are willing to do this soon, we can get it in the version of the function that is released in SciPy 1.12. 

There are other things about documentation that I'd change in all the related function, but I've avoided commenting about those sorts of things here if they were just copied over.

nprav commented:
@mdhaber , should i add the follow up changes in a separate PR?

mdhaber commented:
Yes, thanks!

nprav commented:
@mdhaber 
Working on these comments in: https://github.com/nprav/scipy/tree/cumulative_simpson-clean-up
Will add a PR to the maintenance branch soon.

**Regarding the case of "... elements of `x` identical"**
If all elements of x are identical, it is easy: integral is 0. 

But if only some elements are identical, it is tricky. Eg: x = [0, 0, 1, 1, 2, 2]
This causes ZeroDivisionErrors in the internal functions. 
In Simpson, this is treated by just skipping the ZeroDivisionError points and just taking them as 0: https://github.com/scipy/scipy/blob/main/scipy/integrate/_quadrature.py#L560

We can do the same in cumulative_simpson, but the results may not match with simpson because cumulative_simpson explicitly calculates the intermediate points. More importantly, this does not make much mathematical sense. I think it should be up to the user to determine how to clean/handle their inputs. At the same time, I understand we want to keep the API consistent.

Do you have any thoughts on this?

lucascolley commented:
> Will add a PR to the maintenance branch soon.

All PRs should be to `main`, then the release manager can backport.

nprav commented:
@mdhaber 
Added #19709 for the follow-up comments.

nprav commented on scipy/integrate/_quadrature.py:
```diff
@@ -822,6 +825,260 @@ def simpson(y, *, x=None, dx=1.0, axis=-1, even=_NoValue):
+    x : array_like, optional
+        The coordinate to integrate along. Must be monotonically increasing.
+        If None (default), use spacing `dx` between consecutive elements in 
+        `y`.
```
Added in #19709

nprav commented on scipy/integrate/_quadrature.py:
```diff
@@ -822,6 +825,260 @@ def simpson(y, *, x=None, dx=1.0, axis=-1, even=_NoValue):
+
+    Consider three consecutive points: 
+    :math:`(x_1, y_1), (x_2, y_2), (x_3, y_3)`.
+    where the widths of the 2 subintervals are:
+    :math:`h_1 = x_2 - x_1` and :math:`h_2 = x_3 - x_2`.
+
```
Deleted in #19709
